server/etcdserver/apply/apply_auth.go

```diff
@@ -125,6 +125,11 @@ func (aa *authApplierV3) LeaseRevoke(lc *pb.LeaseRevokeRequest) (*pb.LeaseRevoke
 func (aa *authApplierV3) checkLeasePuts(leaseID lease.LeaseID) error {
 	l := aa.lessor.Lookup(leaseID)
 	if l != nil {
+		// early return for most-common scenario of either disabled auth or admin user
+		if aa.as.IsAdminPermitted(&aa.authInfo) == nil {
```

Comment:
It's not obvious from proposed check that it covers case that auth is disabled. From code design it seems strange  for me that `authApplierV3` will be used at all if auth is disabled.

Reply:
I was surprised too, it even goes through an rw mutex everytime it checks...

https://github.com/etcd-io/etcd/issues/15993#issuecomment-1573550357

I reckon we should discuss a bigger refactoring for the auth system, potentially?

Reply:
Maybe, depends on whether we want to backport this fix. This seems like a trivial but important change. Appliers changed a lot between v3.5 and main.

Reply:
> From code design it seems strange for me that authApplierV3 will be used at all if auth is disabled.

> I reckon we should discuss a bigger refactoring for the auth system, potentially?

Because the existing chain of applier is static. If we want to get rid of authApplier completely when auth isn't enabled, then we need to make the chain dynamic. We don't have to do it, please feel free to evaluate the effort and impact separately if anyone is interested.

> It's not obvious from proposed check that it covers case that auth is disabled.

It's a common "issue" in all existing `(*authStore) IsXXXPermitted(...) error` methods, when auth isn't enabled, they return nil.

> depends on whether we want to backport this fix

I think we need to backport the fix.

Reply:
> we need to make the chain dynamic.  

I haven't researched too deeply yet, but the whole system is already dynamic because the whole auth goes through raft (even enable/disable). I would've expected the enable/disable methods to be a configuration for startup, not a dynamic apply request. Hence the static implementation, I guess.

I'll open a discussion issue up as a feature request, will do some benchmarks in the meantime around the impact of either approach.

Reply:
The link you provide just replaces the whole applier chain with `applierV3Capped` or `applierV3Corrupt`. But you can't change (swap out) part of the chain: `authApplier -> quotaApplier -> backendApplier`. 

It's similar to `string` in golang, it's immutable. You can't modify part of a string; instead, you can only replace the whole string.

Just as I mentioned previously, we don't have to make the applier chain dynamic. But of course, it's open to discussion in a separate session.

Reply:
No doubt on the immutability of the chain, but the pointer to the applier struct needs to be guarded by a mutex when you swap it out from another goroutine?

https://github.com/etcd-io/etcd/blob/release-3.5/server/etcdserver/server.go#L252

let me try to tease the race detector on this case more specifically. Not that it matters much, if you're corrupted/OOS it's game over anyway, it just tickled my inner race detector.